frontend/appflowy_flutter/lib/workspace/application/user/user_workspace_bloc.dart

```diff
@@ -0,0 +1,139 @@
+                  return FlowyResult.failure(e);
+                }),
+              ),
+            );
+          },
+          deleteWorkspace: (workspaceId) async {
+            final result = await _userService.deleteWorkspaceById(workspaceId);
+            emit(
+              state.copyWith(
+                openWorkspaceResult: null,
+                createWorkspaceResult: null,
+                deleteWorkspaceResult:
+                    result.fold((s) => FlowyResult.success(null), (e) {
+                  Log.error(e);
+                  return FlowyResult.failure(e);
+                }),
+              ),
+            );
+          },
+          openWorkspace: (workspaceId) async {
+            final result = await _userService.openWorkspace(workspaceId);
+            emit(
+              state.copyWith(
+                createWorkspaceResult: null,
+                deleteWorkspaceResult: null,
+                openWorkspaceResult:
+                    result.fold((s) => FlowyResult.success(null), (e) {
+                  Log.error(e);
+                  return FlowyResult.failure(e);
+                }),
+              ),
+            );
+          },
+        );
+      },
+    );
+  }
+
+  final UserProfilePB userProfile;
+  final UserBackendService _userService;
+
+  Future<(UserWorkspacePB currentWorkspace, List<UserWorkspacePB> workspaces)?>
+      _fetchWorkspaces() async {
+    final result = await _userService.getCurrentWorkspace();
+    return result.fold((currentWorkspace) async {
+      final result = await _userService.getWorkspaces();
+      return result.fold((workspaces) {
+        return (
+          workspaces.firstWhere(
+            (e) => e.workspaceId == currentWorkspace.id,
+          ),
+          workspaces
+        );
+      }, (e) {
+        Log.error(e);
+        return null;
+      });
+    }, (e) {
+      Log.error(e);
+      return null;
+    });
+  }
+}
+
+@freezed
+class UserWorkspaceEvent with _$UserWorkspaceEvent {
+  const factory UserWorkspaceEvent.initial() = Initial;
+  const factory UserWorkspaceEvent.createWorkspace(String name, String desc) =
+      CreateWorkspace;
+  const factory UserWorkspaceEvent.fetchWorkspaces() = FetchWorkspaces;
+  const factory UserWorkspaceEvent.deleteWorkspace(String workspaceId) =
+      DeleteWorkspace;
+  const factory UserWorkspaceEvent.openWorkspace(String workspaceId) =
+      OpenWorkspace;
+  const factory UserWorkspaceEvent.workspacesReceived(
+    FlowyResult<List<UserWorkspacePB>, FlowyError> workspacesOrFail,
+  ) = WorkspacesReceived;
+}
+
+@freezed
+class UserWorkspaceState with _$UserWorkspaceState {
+  const factory UserWorkspaceState({
+    required UserWorkspacePB? currentWorkspace,
+    required List<UserWorkspacePB> workspaces,
+    required FlowyResult<void, FlowyError>? createWorkspaceResult,
+    required FlowyResult<void, FlowyError>? deleteWorkspaceResult,
+    required FlowyResult<void, FlowyError>? openWorkspaceResult,
```

Comment:
I'm not sure why we need to emit the FlowyResult to the presentation layer?

Reply:
```suggestion
    @Default(null) FlowyResult<void, FlowyError>? createWorkspaceResult,
    @Default(null) FlowyResult<void, FlowyError>? deleteWorkspaceResult,
    @Default(null) FlowyResult<void, FlowyError>? openWorkspaceResult,
```

Reply:
I used them to display the errors. Any other good ideas on the error handling?